### PR TITLE
Implementing the Lyf API support replacing Lydia

### PR DESCRIPTION
**Pull request by citorva**
This merge request made these following changes:
- Remove all references to Lydia and replace the use of the Lydia API in order to recharge the account by the Lyf API
- Update the discord link with a working one
- Replace old icons with a better quality ones

Note: The vendor ID and secret key given in the .env.template file are test API credentials, not the association's credentials.

### CODE REVIEW

File: app/components/hub/leaderboard/Records.tsx [hunk 1]
```diff
@@ -20,6 +20,18 @@ export default function Records() {
     where: { tag: 'leaderboard' }
   });
 
+  if (leaderboard == null) {
+    return (
+      <>
```
Comment (IraSkyx) [inline app/components/hub/leaderboard/Records.tsx:25]:
```suggestion
```

File: app/components/hub/leaderboard/Records.tsx [hunk 2]
```diff
@@ -20,6 +20,18 @@ export default function Records() {
...
+        <TableRow>
+          <TableCell align="center" colSpan={5}>
+            Aucun leaderboard ne peut être affiché
+          </TableCell>
+        </TableRow>
+      </>
```
Comment (IraSkyx) [inline app/components/hub/leaderboard/Records.tsx:31]:
```suggestion
```

File: app/core/utils/topup.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
+import { BinaryLike, KeyObject, createHmac } from 'crypto'
+import base64url from 'base64url'
+
+export function makeMerchantReference(card: number, timestamp: number): string {
```
Comment (IraSkyx) [inline app/core/utils/topup.ts:4]:
```suggestion
export function makeMerchantReference(card: number, timestamp: number) {
```

Comment (IraSkyx) [reply app/core/utils/topup.ts:4]: it's usually better to let TS inference do the work and not explicitly have a return type

File: pages/api/topup/request/index.ts [hunk 4]
```diff
@@ -0,0 +1,5 @@
+import { NextApiRequest, NextApiResponse } from 'next'
```
Comment (IraSkyx) [inline pages/api/topup/request/index.ts:1]: to remove?

File: pages/api/topup/confirm/index.ts [hunk 5]
```diff
@@ -0,0 +1,5 @@
+import { NextApiRequest, NextApiResponse } from 'next'
```
Comment (IraSkyx) [inline pages/api/topup/confirm/index.ts:1]: to remove?

File: app/entities/analytic/queries/getAnalytic.ts [hunk 6]
```diff
@@ -5,5 +5,9 @@ import { resolver } from '@blitzjs/rpc';
 type FindUniqueAnalyticInput = Pick<Prisma.AnalyticFindUniqueArgs, 'where'>;
 
 export default resolver.pipe(resolver.authorize(), async ({ where }: FindUniqueAnalyticInput) => {
-  return await db.analytic.findFirstOrThrow({ where });
+  try {
+    return await db.analytic.findFirstOrThrow({ where });
```
Comment (IraSkyx) [inline app/entities/analytic/queries/getAnalytic.ts:9]:
```suggestion
    return await db.analytic.findFirst({ where });
```

Comment (IraSkyx) [reply app/entities/analytic/queries/getAnalytic.ts:9]: no need for the try/catch

File: app/core/utils/topup.ts [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+export function makeShopOrderReference(card: number, amount: number): string {
+  const card_prefix = card > 0 ? 'p' : 'm'
+
+  return `o${card_prefix}${Math.abs(card)}a${amount}`
+}
+
+export function makeHmac(elements: Array<any>, secret: BinaryLike | KeyObject): string {
+  let encoded = ''
```
Comment (IraSkyx) [inline app/core/utils/topup.ts:17]:
```suggestion
return createHmac('sha1', secret).update(elements.join('*').digest('hex')
```

File: app/core/utils/topup.ts [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+  card: number
+  amount: number
+  reference: string
+  order_reference: string
+  creation_date: number
+  by_credit_card: boolean
+}
+
+export function generateTopUpToken(info: TopUpInfo, secret: BinaryLike | KeyObject): string {
+  let ret = base64url.encode(JSON.stringify(info))
```
Comment (IraSkyx) [inline app/core/utils/topup.ts:41]: do we need this dependency? Wouldn't it work with `Buffer.from(JSON.stringify(info)).toString('base64')`?

File: pages/api/topup/confirm/[token].ts [hunk 9]
```diff
@@ -0,0 +1,136 @@
+import { NextApiRequest, NextApiResponse } from 'next'
+
+import db from 'db'
+import { makeHmac, makeMerchantReference, makeShopOrderReference, parseTopUpToken } from '../../../../app/core/utils/topup'
```
Comment (IraSkyx) [inline pages/api/topup/confirm/[token].ts:4]:
```suggestion
import { makeHmac, makeMerchantReference, makeShopOrderReference, parseTopUpToken } from 'app/core/utils/topup'
```

File: pages/api/topup/confirm/[token].ts [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+
+    const token_info = parseTopUpToken(token, `${process.env.SESSION_SECRET_KEY}`)
+
+    if (token_info == null) {
+      console.error('Jeton de validation mal formé')
+      res.status(404).send('BAD TOKEN')
+      return
+    }
+
+    // Récupération des données
```
Comment (IraSkyx) [inline pages/api/topup/confirm/[token].ts:20]: I'd rather put the comments in english

File: app/core/utils/topup.ts [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+  return `r${card_prefix}${Math.abs(card)}t${Math.ceil(timestamp)}`
+}
+
+export function makeShopOrderReference(card: number, amount: number): string {
+  const card_prefix = card > 0 ? 'p' : 'm'
+
+  return `o${card_prefix}${Math.abs(card)}a${amount}`
+}
+
+export function makeHmac(elements: Array<any>, secret: BinaryLike | KeyObject): string {
```
Comment (IraSkyx) [inline app/core/utils/topup.ts:16]:
```suggestion
export function makeHmac(elements: any[], secret: BinaryLike | KeyObject): string {
```

File: db/schema.prisma [hunk 12]
```diff
@@ -88,6 +93,18 @@ model Transaction {
   updatedAt DateTime @updatedAt
 }
 
+model TopUp {
```
Comment (IraSkyx) [inline db/schema.prisma:96]: I still don't think we need this model, `Transaction` should be sufficient to insert the validated top up

File: pages/api/topup/request/index.ts (hunk 4 above)
Comment (citorva) [reply pages/api/topup/request/index.ts:1]: indeed, I will find a simpler way to handle this path

File: app/core/utils/topup.ts (hunk 8 above)
Comment (citorva) [reply app/core/utils/topup.ts:41]: Base64 is not compatible with the URL format. The base64url is a variation of base64 which provides a string that could be given in an URL

File: pages/api/topup/request/index.ts (hunk 4 above)
Comment (citorva) [reply pages/api/topup/request/index.ts:1]: Finally, I will remove this file, the system work without that

File: app/core/utils/topup.ts (hunk 8 above)
Comment (IraSkyx) [reply app/core/utils/topup.ts:31]: then you can just use `base64url` encoding from NodeJS:

```
const encoded = Buffer.from('Hello world', 'utf8').toString('base64url');
cosnt decoded = Buffer.from(encoded, 'base64').toString('utf8');
```

File: pages/api/topup/request/[userId].ts [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+      console.error(`Tentative rechargement d’une valeur invalide (${amount}) par ${id}`)
+      res.status(400).json({ name: 'Invalid amount' })
+    } else if (method == 'cb' || method == 'lyf') {
+      let by_credit_card = true
+
+      if (method == 'lyf') {
+        by_credit_card = false
+      }
+
+      const user = await db.user.findUnique({ where: { id } })
```
Comment (IraSkyx) [inline pages/api/topup/request/[userId].ts:157]:
```suggestion
      const user = await db.user.findUniqueOrThrow({ where: { id } })
```

File: pages/api/topup/request/[userId].ts [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+
+    const { amount, method } = body
+
+    const id = query.userId as string
+    const tAmount = parseFloat(amount)
+
+    if (Number.isNaN(tAmount) || tAmount <= 0) {
+      console.error(`Tentative rechargement d’une valeur invalide (${amount}) par ${id}`)
+      res.status(400).json({ name: 'Invalid amount' })
+    } else if (method == 'cb' || method == 'lyf') {
```
Comment (IraSkyx) [inline pages/api/topup/request/[userId].ts:150]: Maybe we should use the english version of "cb"

File: pages/api/topup/request/[userId].ts [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+  })
+
+  return ret
+}
+
+function prepare_request(
+  id: string,
+  card: number,
+  amount: number,
+  by_credit_card: boolean
```
Comment (IraSkyx) [inline pages/api/topup/request/[userId].ts:72]:
```suggestion
  byCreditCard: boolean
```

File: pages/api/topup/request/[userId].ts [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+  const shopReference = makeMerchantReference(card, timestamp)
+  const shopOrderReference = makeShopOrderReference(card, tAmount)
+
+  const token = generateTopUpToken(
+    {
+      user_id: id,
+      card,
+      amount: tAmount,
+      reference: shopReference,
+      order_reference: shopOrderReference,
```
Comment (IraSkyx) [inline pages/api/topup/request/[userId].ts:88]:
```suggestion
      orderReference: shopOrderReference,
```

File: pages/api/topup/request/[userId].ts [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+      creation_date: timestamp,
```
Comment (IraSkyx) [inline pages/api/topup/request/[userId].ts:89]:
```suggestion
      creationDate: timestamp,
```

File: pages/api/topup/request/[userId].ts [hunk 18]
```diff
@@ -0,0 +1,180 @@
...
+      additionalData,
+      data.get('enforcedIdentification'),
+    ])
+  }
+
+  return makeHmac(list, `${process.env.LYF_API_SECRET_KEY}`)
+}
+
+function generate_form_get_request(base_url: string, formData: URLSearchParams): string {
+  let ret = base_url
```
Comment (IraSkyx) [inline pages/api/topup/request/[userId].ts:53]:
```suggestion
 return `${base_url}?${formData.toString()}` 
```

Comment (IraSkyx) [reply pages/api/topup/request/[userId].ts:53]: it should already be encoded with this

File: app/core/utils/topup.ts (hunk 8 above)
Comment (citorva) [reply app/core/utils/topup.ts:31]: I did not know

File: app/core/utils/topup.ts [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+  return `r${card_prefix}${Math.abs(card)}t${Math.ceil(timestamp)}`
+}
+
+export function makeShopOrderReference(card: number, amount: number) {
+  const card_prefix = card > 0 ? 'p' : 'm'
+
+  return `o${card_prefix}${Math.abs(card)}a${amount}`
+}
+
+export function makeHmac(elements: Array<any>, secret: BinaryLike | KeyObject): string {
```
[the remark this answers is not in the document]
Comment (citorva) [reply app/core/utils/topup.ts:16]: oups 😅

File: app/core/utils/topup.ts [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+  reference: string
+  order_reference: string
+  creation_date: number
+  by_credit_card: boolean
+}
+
+export function generateTopUpToken(info: TopUpInfo, secret: BinaryLike | KeyObject) {
+  let ret = base64url.encode(JSON.stringify(info))
+
+  return `${ret}.${createHmac('sha1', secret).update(ret).digest('hex')}`
```
[the remark this answers is not in the document]
Comment (citorva) [reply app/core/utils/topup.ts:33]: I don’t understand your suggest

File: app/components/hub/transactions/operations/topUp/TopUp.tsx [hunk 21]
```diff
@@ -1,42 +1,34 @@
...
 
-export type PaymentMethod = 'cb' | 'lydia';
-
 export default function TopUp() {
-  const session = useAuthenticatedSession();
-  const { open, message, severity, onShow, onClose } = useSnackbar();
-  const [paymentMethod, setPaymentMethod] = useState<PaymentMethod>('cb');
+  const { open, message, severity, onClose, onShow } = useSnackbar();
+  const [paymentMethod, setPaymentMethod] = useState<PaymentMethod>('credit');
+  const [topUp, _] = useMutation(requestTopUp);
```
Comment (IraSkyx) [inline app/components/hub/transactions/operations/topUp/TopUp.tsx:16]:
```suggestion
  const [topUp] = useMutation(requestTopUp);
```

File: app/components/hub/transactions/operations/topUp/TopUp.tsx [hunk 22]
```diff
@@ -1,42 +1,34 @@
...
-      .then((info) => window.location.assign(info.mobile_url))
-      .catch((err) => onShow('error', err.message));
+    topUp({
+      amount: data.amount,
+      method: paymentMethod
+    }).then(
+      (url) => {
+        window.location.assign(url as string);
+      },
+      (reason) => {
```
Comment (IraSkyx) [inline app/components/hub/transactions/operations/topUp/TopUp.tsx:28]:
```suggestion
      (error) => {
```

File: app/entities/transactions/mutations/requestTopUp.ts [hunk 23]
```diff
@@ -0,0 +1,135 @@
...
+import { generateTopUpToken, makeHmac, makeMerchantReference, makeShopOrderReference } from 'app/core/utils/topup';
+
+type RequestTopUpInput = {
+  amount: number;
+  method: PaymentMethod;
+};
+
+export type PaymentMethod = 'credit' | 'lyf';
+
+function generate_hmac(data: URLSearchParams, additionalData: string, by_credit_card: boolean): string {
```
Comment (IraSkyx) [inline app/entities/transactions/mutations/requestTopUp.ts:14]:
```suggestion
function generate_hmac(data: URLSearchParams, additionalData: string, byCreditCard: boolean): string {
```

File: app/entities/transactions/mutations/requestTopUp.ts [hunk 24]
```diff
@@ -0,0 +1,135 @@
...
+      data.get('onError'),
+      additionalData,
+      data.get('enforcedIdentification')
+    ]);
+  }
+
+  return makeHmac(list, `${process.env.LYF_API_SECRET_KEY}`);
+}
+
+function prepare_request(id: string, card: number, amount: number, byCreditCard: boolean): URLSearchParams {
```
Comment (IraSkyx) [inline app/entities/transactions/mutations/requestTopUp.ts:55]:
```suggestion
function prepareRequest(id: string, card: number, amount: number, byCreditCard: boolean): URLSearchParams {
```

File: app/entities/transactions/mutations/requestTopUp.ts (hunk 23 above)
Comment (IraSkyx) [inline app/entities/transactions/mutations/requestTopUp.ts:14]:
```suggestion
function generateHMAC(data: URLSearchParams, additionalData: string, byCreditCard: boolean): string {
```

File: app/entities/transactions/mutations/requestTopUp.ts [hunk 25]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+export default resolver.pipe(
+  resolver.authorize(),
+  async (input: RequestTopUpInput, ctx: Ctx) => {
+    if (Number.isNaN(input.amount) || input.amount <= 0 || input.amount >= 1000) {
+      throw new Error('Valeur invalide');
+    }
+
+    let by_credit_card = input.method == 'credit';
```
Comment (IraSkyx) [inline app/entities/transactions/mutations/requestTopUp.ts:127]:
```suggestion
    const byCreditCard = input.method == 'credit';
```

**Review by IraSkyx — APPROVED** [review]

**Pull request merged**